### PR TITLE
Add support for kinds with type restrictions (#203)

### PR DESCRIPTION
**Pull request by Kalin-Rudnicki**

### CODE REVIEW

File: modules/deriving/src/main/scala/shapeless3/deriving/kinds.scala [hunk 1]
```diff
@@ -177,40 +177,44 @@ object K0:
   inline given mkCoproductInstances[F[_], T](using gen: CoproductGeneric[T]): CoproductInstances[F, T] =
     ErasedCoproductInstances[K0.type, F[T], LiftP[F, gen.MirroredElemTypes]](gen): CoproductInstances[F, T]
 
-object K1:
-  type Kind[C, O[_]] = C {
+abstract class K1T[TypeRestriction]:
```
Comment (joroKr21) [inline modules/deriving/src/main/scala/shapeless3/deriving/kinds.scala:180]: If we're doing this we should add a lower bound too. And I wouldn't call this `TypeRestriction` but `LB` and `UB` or `Lower` and `Upper`, something like that.

Comment (Kalin-Rudnicki) [reply modules/deriving/src/main/scala/shapeless3/deriving/kinds.scala:180]: I can make that change!

**Comment by milessabin** [comment]
Now that #206 has been merged I think this belongs in a separate library, or at least a separate module.

In particular I think that adding bounds to the existing `Knn` in this way complicates the typical cases in a significant way (I think the impact on type inference will likely be negative) to support a generalisation which could be easily handled via a separate set of kind definitions.

**Comment by milessabin** [comment]
>  In this case K1 and K11 are exactly as they are before, they are just extending from K1T/K11T, and locking in the bounds.

Possibly I'm being over cautious here. It would be good to test the new definitions against kittens at least.

Aside from that I think the irregularity bothers me. Why no similar treatment for `K2`? Why is the bound on `K11`,
```scala
Kind[C, O[_[_ <: TypeRestriction]]]
```
rather than,
```scala
Kind[C, O[_[t] <: TypeRestriction[t]]]
```
for instance?

In the absence of true kind polymorphism we have to stop somewhere. The current set of kinds is adequate for all of cats (@joroKr21 is that true? I think it is), which makes it a fairly salient place to rest.

**Comment by milessabin** [comment]
> Actually I think it should be:
> ```
> Kind[C, O[_] >: LB <: UB]
> ```
> So that LB and UB can be either type lambdas or whatever is needed

Agreed that both upper and lower bounds would be needed for symmetry, however that's missing the point I was making, which was that a bound on a type of kind `* -> *` could be another type of kind `* -> *` rather than a type of kind `*`.

There is literally no end to this, which I why I chose to draw the line where I did.

@joroKr21 you're the primary maintainer now, so this is your call, but I think you'll be making your life difficult if you go down this road.

**Comment by joroKr21** [comment]
> Agreed that both upper and lower bounds would be needed for symmetry, however that's missing the point I was making, which was that a bound on a type of kind * -> * could be another type of kind * -> * rather than a type of kind *.

I think it might work with type lambdas in Scala 3 but if I'm wrong about that then I agree it's not worth the hassle. We have to try first

**Comment by milessabin** [comment]
> I think it might work with type lambdas in Scala 3

I think the type variable is essential ... `T[t] <: U[t]` isn't the same bound as `T[_] <: U[_]` because the second `_` is interpreted as implicitly quantified, ie. more or less equivalent to `T[_] <: U[Any]` where `U` is covariant.

Which raises another issue ... what about variance? Each variance annotation creates a new kind.

**Comment by joroKr21** [comment]
Ah yeah it doesn't work as I expected:
```scala
scala> class Foo[F[_] <: [x] =>> List[x]]
// defined class Foo

scala> new Foo[List]
-- Error: ----------------------------------------------------------------------
1 |new Foo[List]
  |        ^^^^
  | Type argument List does not have the same kind as its bound [_$1] <: List
1 error found

scala> class Foo[F <: [x] =>> List[x]]
// defined class Foo

scala> new Foo[List]
val res0: Foo[List] = Foo@50df37eb
```

**Pull request closed without merging**